platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/migratable/FeatureProcessor.java

```diff
@@ -316,71 +420,141 @@ public void processUninstalledFeature(
-   * then be finally started as it was on the original system. Otherwise, if it is not in the
-   * started state, it will simply be started.
+   * Processes the specified feature for starting.
+   *
+   * <p><i>Note:</i> If the feature was uninstalled, it will be installed and then later started to
+   * get to the started state. The change from installed to started will require another process
+   * attempt as we only want to deal with one state change at a time. The next processing round will
+   * see the state of the feature in memory as installed instead of uninstalled like it is right
+   * such that it can be finally started as it was on the original system. Otherwise, if it is not
+   * in the started state, it will simply be started.
    *
+   * @param jfeature the original feature information
    * @param feature the current feature from memory
-   * @param id the feature id
-   * @param state the feature state in memory
-   * @param region the region for the feature
    * @param tasks the task list where to record tasks to be executed
    */
-  public void processStartedFeature(
-      Feature feature, String id, FeatureState state, String region, TaskList tasks) {
+  public void processStartedFeatureToRecordTasks(
+      JsonFeature jfeature, Feature feature, TaskList tasks) {
+    final String id = feature.getId();
+    final FeatureState state = service.getState(id);
+
     if (state == FeatureState.Uninstalled) {
       // we need to first install it and on the next round, start it
-      // as such, let's make sure to increase the number of attempts left for the start operation
-      tasks.add(Operation.INSTALL, id, r -> installFeature(r, feature));
-      tasks.increaseAttemptsFor(Operation.START);
+      addCompoundInstallTaskFor(jfeature, tasks);
+      return;
     } else if (state != FeatureState.Started) {
-      tasks.add(Operation.START, id, r -> startFeature(r, feature, region));
+      // should use feature's region but we cannot figure out how to get it
+      tasks.add(Operation.START, id, r -> startFeature(r, feature, jfeature.getRegion()));
+    }
+    if (jfeature.isRequired() != service.isRequired(feature)) {
+      addCompoundUpdateTaskFor(jfeature, feature, tasks);
+    }
+  }
+
+  private void addCompoundInstallTaskFor(JsonFeature jfeature, TaskList tasks) {
+    // we shall verify if the installed feature is in the proper required state on a subsequent pass
+    // and if not, update its requirements appropriately
+    tasks.add(
+        Operation.INSTALL,
+        jfeature.getId(),
+        HashMap<String, Set<JsonFeature>>::new,
+        (jfeatures, r) -> installFeatures(r, jfeatures),
+        (id, jfeatures) ->
+            jfeatures.computeIfAbsent(jfeature.getRegion(), r -> new HashSet<>()).add(jfeature));
+  }
+
+  private void addCompoundUninstallTaskFor(Feature feature, TaskList tasks) {
+    // since we do not know how to get the region for a feature yet, use the default one: ROOT
```

Comment:
By subsystem, you mean region I guess?

Based on what I saw in the current Karaf API, this is what happens anyway since they don't properly support in all areas the notion of region. As stated in the code, there is no way to know which region a feature is installed in. They have also hardcoded checking only the root region for attributes such as installed, is-required, ...

The current logic gets us closer to supporting it all once some of these shortcomings are addressed in the future. Also the goal of this migratable is to restore the system in the exact same state it was. Making decisions to change things would be part of upgrading from version X to Y.